src/main/java/reposense/model/RepoLocation.java

```diff
@@ -1,42 +1,32 @@
 package reposense.model;
 
-import java.net.MalformedURLException;
-import java.net.URL;
-import java.nio.file.Files;
-import java.nio.file.InvalidPathException;
-import java.nio.file.Path;
-import java.nio.file.Paths;
-import java.util.regex.Matcher;
-import java.util.regex.Pattern;
+import java.util.Optional;
 
 import reposense.parser.InvalidLocationException;
+import reposense.parser.RepoLocationParser;
 
 /**
  * Represents a repository location.
  */
 public class RepoLocation {
-    private static final String GIT_LINK_SUFFIX = ".git";
-    private static final Pattern GIT_REPOSITORY_LOCATION_PATTERN =
-            Pattern.compile("^.*github.com\\/(?<org>.+?)\\/(?<repoName>.+?)\\.git$");
-
     private final String location;
+    private final transient Optional<String> parsedBranch;
     private final String repoName;
     private String organization;
 
     /**
      * @throws InvalidLocationException if {@code location} cannot be represented by a {@code URL} or {@code Path}.
      */
     public RepoLocation(String location) throws InvalidLocationException {
-        verifyLocation(location);
-        this.location = location;
-        Matcher matcher = GIT_REPOSITORY_LOCATION_PATTERN.matcher(location);
+        this(RepoLocationParser.parse(location));
+    }
 
-        if (matcher.matches()) {
-            organization = matcher.group("org");
-            repoName = matcher.group("repoName");
-        } else {
-            repoName = Paths.get(location).getFileName().toString().replace(GIT_LINK_SUFFIX, "");
-        }
+    private RepoLocation(String[] repoLocationDetails) {
+        assert repoLocationDetails.length == 4;
+        location = repoLocationDetails[RepoLocationParser.LOCATION_INDEX];
```

Comment:
I did not do this initially, as I thought such a change may be out of scope of the PR.
However, this may help to catch bugs in the implementation of this PR.